src/plugins/ucx/ucx_backend.cpp

```diff
@@ -483,81 +515,94 @@ void nixlUcxEngine::progressFunc()
         NIXL_PERROR << "write to progress thread control pipe failed";
     pthr.join();
 }
 
-void nixlUcxEngine::progressThreadRestart()
+int nixlUcxThreadEngine::vramApplyCtx()
 {
     progressThreadStop();
     progressThreadStart();
+    return nixlUcxEngine::vramApplyCtx();
 }
 
-/****************************************
- * Constructor/Destructor
-*****************************************/
+void nixlUcxThreadEngine::appendNotif(std::string remote_name, std::string msg)
+{
+    if (isProgressThread()) {
+        /* Append to the private list to allow batching */
+        std::unique_lock<std::mutex> lock(pthrActiveLock);
+        notifPthrPriv.push_back(std::make_pair(std::move(remote_name), std::move(msg)));
+    } else {
+        nixlUcxEngine::appendNotif(std::move(remote_name), std::move(msg));
+    }
+}
 
-nixlUcxEngine::nixlUcxEngine(const nixlBackendInitParams *init_params)
-    : nixlBackendEngine(init_params),
-      pthrControlPipe{0, 0} {
-    size_t numWorkers;
-    std::vector<std::string> devs; /* Empty vector */
-    nixl_b_params_t* custom_params = init_params->customParams;
+void nixlUcxThreadEngine::notifProgressCombineHelper(notif_list_t &src, notif_list_t &tgt)
+{
+    const std::lock_guard<std::mutex> lock(notifMtx);
+    moveNotifList(src, tgt);
+}
 
-    if (init_params->enableProgTh) {
-        if (!nixlUcxMtLevelIsSupported(nixl_ucx_mt_t::WORKER)) {
-            throw std::invalid_argument("UCX library does not support multi-threading");
-        }
+void nixlUcxThreadEngine::notifProgress()
+{
+    notifProgressCombineHelper(notifPthrPriv, notifPthr);
+}
 
-        if (pipe(pthrControlPipe) < 0) {
-            throw std::runtime_error("Couldn't create progress thread control pipe");
-        }
+nixl_status_t nixlUcxThreadEngine::getNotifs(notif_list_t &notif_list)
+{
+    if (!notif_list.empty())
+        return NIXL_ERR_INVALID_PARAM;
 
-        // This will ensure that the resulting delay is at least 1ms and fits into int in order for
-        // it to be compatible with poll()
-        pthrDelay = std::chrono::ceil<std::chrono::milliseconds>(
-            std::chrono::microseconds(init_params->pthrDelay < std::numeric_limits<int>::max() ?
-                                          init_params->pthrDelay :
-                                          std::numeric_limits<int>::max()));
-        pthrOn = true;
+    moveNotifList(notifMainList, notif_list);
+    notifProgressCombineHelper(notifPthr, notif_list);
+    return NIXL_SUCCESS;
+}
+
+/****************************************
+ * Constructor/Destructor
+*****************************************/
+
+std::unique_ptr<nixlUcxEngine>
+nixlUcxEngine::create(const nixlBackendInitParams &init_params)
+{
+    nixlUcxEngine *engine;
+    if (init_params.enableProgTh) {
+        engine = new nixlUcxThreadEngine(init_params);
     } else {
-        pthrOn = false;
+        engine = new nixlUcxEngine(init_params);
     }
+    return std::unique_ptr<nixlUcxEngine>(engine);
```

Comment:
```C++
if (init_params.enableProgTh) {
    return std::make_unique<nixlUcxThreadEngine>(init_params);
}
return std::make_unique<nixlUcxEngine>(init_params);

Reply:
Yes, so I suggested to make ctor it public, but disable it from the usage via protected structure:
```C++
class nixlUcxEngine {
public:
    nixlUcxEngine(const nixlBackendInitParams &, Accessor);
    static std::unique_ptr<nixlUcxEngine> create(const nixlBackendInitParams &);
protected:
    struct Accessor {};
    ...
};
```

The same approach is used in the example in this topic https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c50-use-a-factory-function-if-you-need-virtual-behavior-during-initialization